### ISSUE TITLE
Errors when creating WireGuard interface with NDB

### ISSUE THREAD

**Issue by qingbozhou** [svinota/pyroute2#922]
Hi I run the following code with pyroute2 0.6.9, on Ubuntu 20.04, during server startup:

```
ndb = NDB()
interface = ndb.interfaces.create(kind="wireguard", ifname="wg0")
interface.set("state", "up")
interface.set("mtu", 1420)
interface.add_ip("192.168.1.1/24")
interface.commit()
ndb.close()
```

sometimes I see errors like this:

```
WARNING:pr2modules.ndb.main:unsupported event ignored: <class 'pr2modules.netlink.nlmsg'>
ERROR:pyroute2.ndb.140006786322976.main:exception <(22, 'Invalid argument')> in source localhost
ERROR:pyroute2.ndb.140006786322976.sources.localhost:source error: <class 'queue.Full'>
```

The warning on first line can be ignored and I think the latest version already suppresses this warning. Could you please help explain the 2 errors? 

Sometimes the interface is successfully created but sometimes the process seems stuck, and `wg show` only outputs:

```
interface: wg0
  listening port: 12676
```

I'm thinking of several ways to diagnose or mitigate the issue:
* Upgrading pyroute2
* More logging to better understand when the errors are emitted or the process is stuck.
* Use IPDB - although deprecated, will it be free of such errors?
* Delayed running - do the errors have anything to do with server startup?

**Comment by qingbozhou** [comment]
Another run resulted in the following slightly different error messages:
```
WARNING:pr2modules.ndb.main:unsupported event ignored: <class 'pr2modules.netlink.nlmsg'>
ERROR:pyroute2.ndb.140222385738272.main:exception <(22, 'Invalid argument')> in source localhost
WARNING:pr2modules.ndb.main:unsupported event ignored: <class 'pr2modules.netlink.nlmsg'>
ERROR:pyroute2.ndb.140222385738272.main:exception <(22, 'Invalid argument')> in source localhost
```

This time I checked with `ip address show wg0` and the interface already has IP address, so the process might be stuck at `ndb.close()`?

**Comment by svinota** [comment]
Thanks for the report, will be investigated today

**Comment by svinota** [comment]
```
ERROR:pyroute2.ndb.140006786322976.main:exception <(22, 'Invalid argument')> in source localhost
```
Looks like a netlink API call returned errno 22, EINVAL. Which one — hard to tell without the debug log. To turn on the debug logging:
```python
# to the console
NDB(log='debug')

# to a file
import logging
NDB(log=('file_name.log', logging.DEBUG))
```
It is very chatty and (beware!) may contain some sensitive information about your network topology, so pls review carefully if you decide to post it here.

```
ERROR:pyroute2.ndb.140006786322976.sources.localhost:source error: <class 'queue.Full'>
```

The only non blocking queue put in this data path is here: https://github.com/svinota/pyroute2/blob/960bb0928e3e5fd41fc16ddc99da64d5f40e7cb7/pyroute2.core/pr2modules/netlink/nlsocket.py#L586-L603

And the reason to use non blocking put here is that we have no choice but to timely consume all the data from the netlink socket, otherwise netlink session will not guarantee the data integrity due to possible missed messages, and all the subsequent reads from the socket will return error.

Now the important part: the buffer queue is limited by default with 4096 blocks × 64000 bytes each (nlsocket module uses bytearrays to fetch the data into), every message pack contains less than 64K of data, this is the upper limit of reading block). That must be enough for most the cases, but obviously your server hit the limit.

To mitigate that on the current codebase:
```python
from pyroute2 import config, NDB

# run this before any NDB or IPRoute instantiation
config.async_qsize = x  # here try to find a reasonable limit but more than the default 4096

NDB(log=('file_name.log', logging.DEBUG))
```

**Comment by svinota** [comment]
> This time I checked with ip address show wg0 and the interface already has IP address, so the process might be stuck at ndb.close()?

I would love to see the debug log for this case, as it looks like a bug, but see above — be careful not to disclose any sensitive info posting the debug log fragments here.

**Comment by svinota** [comment]
TODO: update the API call exception to contain more info, so there will be less need in the debug log in relatively simple cases.

**Comment by qingbozhou** [comment]
Thanks for the quick and detailed response. I'll add debug logging (probably next week) and see if we can see anything interesting.

**Comment by svinota** [comment]
> Something equivalent to `ip link add dev wg0 type wireguard` but would be great in pure Python

Basically it is:

```python
from pyroute2 import IPRoute

with IPRoute() as ipr:
    ipr.link('add', ifname='wg0', kind='wireguard')
```

But be careful — it does not ensure the interface to exist in the system, the call only triggers the interface creation, and immediately returns. There is a big chance that the next line of code will be executed before the interface appears in the system. Actually, it's exactly the same what `ip link add` does — it only sends a request via netlink to the kernel.

To be sure that the link exists, one can use something like that (writing out of the head, so there may be typos, but the idea is like that):
```python
import time
from typing import List
from pyroute2 import IPRoute

def ensure_inteface(
    ipr: IPRoute,
    ifname: str,
    max_tries: int = 5,
    sleep: float = 0.2
) -> List[int]:
    '''ensure interface is created.

    raises: NetlinkError, TimeoutError
    '''
    for _ in range(max_tries):
        # returns a list, but this list must be accepted wherever IPRoute expects an interface index
        if wg_index := ipr.link_lookup(ifname=ifname):
            return wg_index
        time.sleep(sleep)
    raise TimeoutError('wireguard interface creation error')

with IPRoute() as ipr:
    ipr.link('add', ifname='wg0', kind='wireguard')
    wg_index = ensure_interface(ipr, 'wg0')
```

**Comment by qingbozhou** [comment]
Yeah I saw the async nature of IPRoute documented in https://docs.pyroute2.org/ipdb.html#ipdb-ndb-iproute

Thanks for the sample code!
